### PR TITLE
Resolve NestedSlot detection and https://github.com/ArcBees/gwtp-idea-plugin/issues/59

### PR DESCRIPTION
**Pull request by BenDol**
Resolves:
- https://github.com/ArcBees/gwtp-idea-plugin/issues/59
- https://github.com/ArcBees/gwtp-idea-plugin/issues/74

Tested and is working on my fork repository. I removed the need for maven on my fork repository so I could use hotswapping and debugging which i wasn't sure how to do with this strange maven implementation. Thanks!

### CODE REVIEW

**Comment by BenDol** [comment]
Sorry not able to test compilation since I'm cherry picking on a project that isn't setup to compile. Should pass the TC build now :+1:

**Comment by meriouma** [comment]
Thanks for the contribution!

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 1]
```diff
@@ -17,15 +17,15 @@
 package com.arcbees.plugin.idea.dialogs;
 
 import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.module.Module;
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.ui.DialogWrapper;
-import com.intellij.psi.JavaPsiFacade;
-import com.intellij.psi.PsiClass;
-import com.intellij.psi.PsiMember;
+import com.intellij.psi.*;
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: Revert the star import

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 2]
```diff
@@ -90,28 +92,46 @@ protected JComponent createCenterPanel() {
...
+        PsiClass nestedSlotClass = JavaPsiFacade.getInstance(project)
+            .findClass("com.gwtplatform.mvp.client.presenter.slots.NestedSlot", scope);
 
         slots = new HashMap<String, PsiMember>();
-        Query<PsiMember> query = AnnotatedMembersSearch.search(psiClass, GlobalSearchScope.allScope(project));
-        query.forEach(new Processor<PsiMember>() {
+
+        AnnotatedMembersSearch.search(contentSlotClass,
+            GlobalSearchScope.allScope(project)).forEach(new Processor<PsiMember>() {
             public boolean process(PsiMember psiMember) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: I think the alignment is off here

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 3]
```diff
@@ -90,28 +92,46 @@ protected JComponent createCenterPanel() {
...
                 slots.put(getSlot(psiMember), psiMember);
+                return true;
+            }
+        });
 
+        ReferencesSearch.search(nestedSlotClass,
+            GlobalSearchScope.moduleScope(module), false).forEach(new Processor<PsiReference>() {
+            public boolean process(PsiReference psiReference) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: I think the alignment is off here too

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 4]
```diff
@@ -90,28 +92,46 @@ protected JComponent createCenterPanel() {
...
+                if(psiReference instanceof PsiQualifiedReferenceElement) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: space after `if`

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 5]
```diff
@@ -90,28 +92,46 @@ protected JComponent createCenterPanel() {
...
+                    PsiElement parent = ((PsiQualifiedReferenceElement) psiReference).getParent();
+                    if(parent instanceof PsiNewExpression) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: space after `if`

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 6]
```diff
@@ -90,28 +92,46 @@ protected JComponent createCenterPanel() {
...
+                        PsiElement field = parent.getParent();
+                        if(field instanceof PsiField) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: space after `if`

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java [hunk 7]
```diff
@@ -120,4 +140,4 @@ private String getSlot(PsiMember psiMember) {
 
         return classname + "." + psiMember.getName();
     }
-}
+}
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: Keep the newline at end of file

File: src/main/java/com/arcbees/plugin/idea/utils/PackageHierarchy.java [hunk 8]
```diff
@@ -274,37 +285,40 @@ public void run() {
...
-                // add subpackages
-                final PsiDirectory[] subdirectories = directory.getSubdirectories();
-                for (PsiDirectory subdirectory : subdirectories) {
-                    final PsiPackage aPackage = JavaDirectoryService.getInstance().getPackage(subdirectory);
-                    if (aPackage != null && !PackageUtil.isPackageDefault(aPackage)) {
-
-                        PackageRoot packageRoot = new PackageRoot(root, aPackage);
-                        topLevelPackages.add(packageRoot);
+        Module[] modules = ModuleManager.getInstance(myProject).getModules();
+        for(Module module : modules) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageHierarchy.java]: space after `for`

File: src/main/java/com/arcbees/plugin/idea/utils/PackageHierarchy.java [hunk 9]
```diff
@@ -274,37 +285,40 @@ public void run() {
...
+            final ModuleRootManager rootManager = ModuleRootManager.getInstance(module);
+
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageHierarchy.java]: Remove this line

File: src/main/java/com/arcbees/plugin/idea/utils/PackageHierarchy.java [hunk 10]
```diff
@@ -274,37 +285,40 @@ public void run() {
...
+                    continue;
+                }
+                final PsiPackage directoryPackage = JavaDirectoryService.getInstance().getPackage(directory);
+                if (directoryPackage == null || PackageUtil.isPackageDefault(directoryPackage)) {
+                    // add subpackages
+                    final PsiDirectory[] subdirectories = directory.getSubdirectories();
+                    for (PsiDirectory subdirectory : subdirectories) {
+                        final PsiPackage aPackage = JavaDirectoryService.getInstance().getPackage(subdirectory);
+                        if (aPackage != null && !PackageUtil.isPackageDefault(aPackage)) {
+
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageHierarchy.java]: remove newline

File: src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package java.com.arcbees.plugin.idea.utils;
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java:17]: Add 2015 Arcbees copyright

File: src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        PackageIndex that = (PackageIndex) o;
+
+        if (module != null ? !module.equals(that.module) : that.module != null) return false;
+        return !(name != null ? !name.equals(that.name) : that.name != null);
+
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java]: remove newline

File: src/main/java/com/arcbees/plugin/idea/wizards/createpresenter/CreatePresenterAction.java [hunk 13]
```diff
@@ -123,7 +124,13 @@ public void beforeActionPerformedUpdate(@NotNull AnActionEvent e) {
     @Override
     public void actionPerformed(AnActionEvent e) {
         project = e.getProject();
-        presenterConfigModel = new PresenterConfigModel(project);
+        module = e.getData(LangDataKeys.MODULE);
+        if(module == null) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/plugin/idea/wizards/createpresenter/CreatePresenterAction.java]: space after `if`

**Comment by meriouma** [comment]
Few code style issues, otherwise it looks good

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java (hunk 1 above)
Comment (BenDol) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: Thanks Intellij import format

File: src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright 2013 ArcBees Inc.
```
Comment (olafleur) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java:2]: 2015

Comment (BenDol) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java:2]: Ah, just copy and pasted from the other files.

Comment (olafleur) [inline src/main/java/com/arcbees/plugin/idea/utils/PackageIndex.java:2]: Yeah, I understand, no problem. We try to have the inception year of the class as the copyright.

File: src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java (hunk 1 above)
Comment (olafleur) [inline src/main/java/com/arcbees/plugin/idea/dialogs/ContentSlotDialog.java]: You can set that up in IntelliJ to wait until you have 999 classes to put the star (which means in practice, never).

**Pull request merged**